File: keras/mnist_loader.py
# 2019/12/30 Hiroyuki Ogasawara
# vim:ts=4 sw=4 et:

import os
import struct
import numpy as np

class MNistLoader:

    def __init__( self, data_path ):
        self.mnist_data= []
        self.load( data_path )

    def load( self, data_path ):
        mnist_files= [
            'train-images-idx3-ubyte',
            't10k-images-idx3-ubyte',
            'train-labels-idx1-ubyte',
            't10k-labels-idx1-ubyte',
        ];
        for name in mnist_files:
            full_path= os.path.join( data_path, name )
            with open( full_path, 'rb' ) as fi:
                self.mnist_data.append( fi.read() )

    def byteToFloat( self, data ):
        float_data= np.zeros( 28*28, dtype='float' )
        for i,b in enumerate(data):
            float_data[i]= b/255.0
        return  float_data

    def getImage( self, index, data_index ):
        image= self.mnist_data[data_index]
        base= 16+28*28*index
        return  self.byteToFloat( image[base:base+28*28] )

    def getFloatImage( self, index ):
        return  self.getImage( index, 0 )

    def getFloatImageTest( self, index ):
        return  self.getImage( index, 1 )

    def getLabel( self, index ):
        label= self.mnist_data[2]
        return  int(label[8 + index])

    def getLabelTest( self, index ):
        label= self.mnist_data[3]
        return  int(label[8 + index])

    def getFloatLabel10( self, index ):
        label= np.zeros( 10, dtype='float' )
        label[self.getLabel(index)]= 1.0
        return  label

    def getFloatLabel10Test( self, index ):
        label= np.zeros( 10, dtype='float' )
        label[self.getLabelTest(index)]= 1.0
        return  label

    def getAll( self ):
        x_train= np.ndarray( shape=(60000,1,28,28), dtype='float' )
        y_train= np.ndarray( shape=(60000,10), dtype='float' )
        for i in range(60000):
            x_train[i]= self.getFloatImage( i ).reshape(1,28,28)
            y_train[i]= self.getFloatLabel10( i )
        x_test= np.ndarray( shape=(10000,1,28,28), dtype='float' )
        y_test= np.ndarray( shape=(10000,10), dtype='float' )
        for i in range(10000):
            x_test[i]= self.getFloatImageTest( i ).reshape(1,28,28)
            y_test[i]= self.getFloatLabel10Test( i )
        return  (x_train,y_train),(x_test,y_test)


def print28( data ):
    for y in data:
        a= ''
        for x in y:
            if x > 0.5:
                a+='#'
            else:
                a+='.'
        print( a )


def main():
    loader= MNistLoader( '../mnist' )
    for i in range(10):
        index= i
        print( loader.getLabel( index ) )
        print( loader.getFloatLabel10( index ) )
        print28( loader.getFloatImage( index ).reshape(28,28) )


if __name__=='__main__':
    main()

File: pytorch/mnist_loader.cpp
// 2018/09/10 Hiroyuki Ogasawara
// vim:ts=4 sw=4 noet:

#include	"mnist_loader.h"
#include	<iostream>
#include	<fstream>
#include	<assert.h>
#include	<string.h>


template<typename T>
void zero_free( T*& ptr )
{
	if( ptr ){
		free( ptr );
		ptr= nullptr;
	}
}


MNistLoader::MNistLoader( const char* mnist_data_path )
{
	LoadImageFile( mnist_data_path );
}


MNistLoader::~MNistLoader()
{
	Finalize();
}


void	MNistLoader::Finalize()
{
	for( void* ptr : ImageList ){
		zero_free( ptr );
	}
	ImageList.clear();
}


static void*	LoadFileData( const char* file_name )
{
	std::ifstream	file( file_name, std::ios::binary|std::ios::ate );
	if( file ){
		size_t	size= file.tellg();
		void*	ptr= malloc( size );
		file.seekg( 0, std::ios_base::beg );
		file.read( (char*)ptr, size );
		return	ptr;
	}
	std::cerr << file_name << " not found\n";
	assert( 0 );
	return	nullptr;
}


void	MNistLoader::LoadImageFile( const char* mnist_data_path )
{
	Finalize();
	const char*	file_table[]= {
		"/train-images-idx3-ubyte",
		"/t10k-images-idx3-ubyte",
		"/train-labels-idx1-ubyte",
		"/t10k-labels-idx1-ubyte",
	};
	std::string	data_path= mnist_data_path;
	for( const auto* file_name : file_table ){
		auto	path= data_path + file_name;
		ImageList.push_back( LoadFileData( path.c_str() ) );
	}
}


template<typename ST>
auto	GetData( void* ptr, int index )
{
	return	reinterpret_cast<ST*>( ptr )->Get( index );
}


const uint8_t*	MNistLoader::GetImage( unsigned int index ) const
{
	assert( index < TRAINDATA_SIZE );
	return	GetData<ImageDataHeader>( ImageList[0], index );
}


const uint8_t*	MNistLoader::GetImageTest( unsigned int index ) const
{
	assert( index < TESTDATA_SIZE );
	return	GetData<ImageDataHeader>( ImageList[1], index );
}


const uint8_t	MNistLoader::GetLabel( unsigned int index ) const
{
	assert( index < TRAINDATA_SIZE );
	return	GetData<LabelDataHeader>( ImageList[2], index );
}


const uint8_t	MNistLoader::GetLabelTest( unsigned int index ) const
{
	assert( index < TESTDATA_SIZE );
	return	GetData<LabelDataHeader>( ImageList[3], index );
}


void	MNistLoader::GetFloatImage_( float* buffer, const uint8_t* image ) const
{
	auto*	ptr= image;
	float*	fptr= buffer;
	for( unsigned int pi= 0 ; pi< IMAGE_PIXSIZE ; pi++ ){
		*fptr++= *ptr++ * (1.0f/255.0f);
	}
}


void	MNistLoader::GetFloatLabel_( float* buffer, uint8_t label ) const
{
	assert( label < LABEL_SIZE );
	memset( buffer, 0, sizeof(float) * LABEL_SIZE );
	buffer[label]= 1.0f;
}


void	MNistLoader::GetFloatImage( float* buffer, unsigned int index ) const
{
	GetFloatImage_( buffer, GetImage( index ) );
}


void	MNistLoader::GetFloatLabel10( float* buffer, unsigned int index ) const
{
	GetFloatLabel_( buffer, GetLabel( index ) );
}


void	MNistLoader::GetFloatImageTest( float* buffer, unsigned int index ) const
{
	GetFloatImage_( buffer, GetImageTest( index ) );
}


